### ISSUE TITLE
Add an async EventHook class to hook global events

### ISSUE THREAD

**Issue by AsafFisher** [rust-windowing/winit#1838]
Hi, I'd like to add an `EventHook` utils to this project. Is there anyone against it?
It will help people to get global events if needed for applications such as remote control apps etc...

Usage:
```rust
match EventHook::new(
     vec![EventType::MouseMoved],
     |_a, _b, d| {
         println!("{:?}", d.location());
         None
     },
 ) {
     Ok(hook) => unsafe {
         loop{
              async::sleep();
         }
     },
     Err(_) => (assert!(false)),
 }
```
I noticed that there are no bindings to the windows hook system (aka `SetWindowsHookExA`) equivalent on macOS (`CGCreateEventTap`)  so I added [bindings](https://github.com/servo/core-foundation-rs/pull/436) to them on [core-foundation-rs](https://github.com/servo/core-foundation-rs)

Will be happy to hear what you all think and improvement suggestions.

**Comment by maroider** [comment]
Could you expand a bit on why you'd need "global input" to implement something like a remote desktop client application? Is there anything about the currently available events that is insufficient for such an application?

I guess this is related to #1821, which you opened a couple of weeks ago. The following goes for both issues: What _concrete_ problem would this solve that device events can't?

**Comment by AsafFisher** [comment]
Yes, the current events are always relative (on some platforms) to an `NSApp`. This means that while the app is in the background, It will not receive the mouse position.
As mentioned here: 
_Note: Applications normally receive events from the keyboard and mouse only when they are in the foreground (that is, active). Although applications running in the background do not normally receive key and mouse events, low-level mechanisms called event taps make it possible for a background application to receive events and act upon them._


 When I think about it the example of Remote Control platform is not a good example, but I am about to do a project where I need to know the location of the mouse in the background.

**Comment by chrisduerr** [comment]
> When I think about it the example of Remote Control platform is not a good example, but I am about to do a project where I need to know the location of the mouse in the background.

I'd be curious what that project is. It might be possible on some platforms, but certainly not on all of them. So implementing this uniformly in winit is impossible.

Either way, even if it was implemented it should be integrated into the event loop.

**Comment by AsafFisher** [comment]
> > When I think about it the example of Remote Control platform is not a good example, but I am about to do a project where I need to know the location of the mouse in the background.
> 
> I'd be curious what that project is. It might be possible on some platforms, but certainly not on all of them. So implementing this uniformly in winit is impossible.
> 

What is an example of a platform that will not support it?
> Either way, even if it was implemented it should be integrated into the event loop.

I don't think its a part of the regular `event_loop`. It is as a matter of a fact not even a loop.

I am making a cross platform remote control utility. When I will have a prototype of it, I will be more then happy to share :smile:

**Comment by chrisduerr** [comment]
> What is an example of a platform that would not support it? If you are talking about web assembly I think it is alright that it will not work there, I will make sure no mistakes will accrue.

Wayland.

> I don't think its a part of the regular event_loop. It is as a matter of a fact not even a loop. It integrates with an existing global loop.

What the operating system uses to dispatch these events should not matter. Events emitted from winit should be distributed uniformly.

**Comment by AsafFisher** [comment]
On linux I will implement it on the `/dev/input/event` level.

**Comment by maroider** [comment]
I'm still not quite sure what it is that you're making. Is it something like [barrier]/[synergy]?

> On linux I will implement it on the `/dev/input/event` level. Yes, it will require higher privilege to run, which is fine.

I feel like we're getting into "doesn't quite make sense for Winit" territory.

[barrier]: https://github.com/debauchee/barrier
[synergy]: https://symless.com/

**Comment by AsafFisher** [comment]
I mean I don't think it will harm to have this functionality... I think splitting libraries too much is not that good. I want to add functionality to the event system of winit. On most platforms it's part of the event system.

**Comment by msiglreith** [comment]
Is this something which could be covered by the existing `DeviceEvent` API?
(only implemented for a few platforms right now afaik)

**Comment by AsafFisher** [comment]
I don't really understand your question

**Comment by chrisduerr** [comment]
The `DeviceEvent` API already sends global mouse events disconnected from the window on some platforms like X11.

**Comment by AsafFisher** [comment]
But the functionality on most platforms is not a "DeviceEvent" Its a hook on the Event system.
